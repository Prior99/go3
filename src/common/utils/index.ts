export * from "./board-color";
export * from "./board-sizes";
export * from "./encrypt";
export * from "./environment";
export * from "./is-browser";
export * from "./rating";
export * from "./rank";
export * from "./authorization";
