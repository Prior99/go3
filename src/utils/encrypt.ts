import { createHmac } from "crypto";

const SECRET = "Eech3ahm";

export function hash(value: string) {
    return createHmac("sha256", SECRET).update(value).digest("hex");
}
