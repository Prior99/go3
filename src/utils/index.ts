export * from "./environment";
export * from "./require-login";
