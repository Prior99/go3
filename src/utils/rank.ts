export enum RankClass {
    KYU = "kyu",
    DAN = "dan",
}

export class Rank {
    public rank: number;
    public rankClass: RankClass;
    public rating: number;

    constructor(rating: number) {
        this.rating = rating;
        if (rating < 2100) {
            this.rank = Math.round((2000 - rating) / 100 + 1);
            this.rankClass = RankClass.KYU;
        } else {
            this.rank = Math.round((rating - 2000) / 100);
            this.rankClass = RankClass.DAN;
        }
    }

    public format() {
        if (this.rankClass === RankClass.KYU) {
            return `${this.rank} Kyu`;
        }
        return `${this.rank} Dan`;
    }

    public formatShort() {
        if (this.rankClass === RankClass.KYU) {
            return `${this.rank}k`;
        }
        return `${this.rank}d`;
    }

    public equals(other: Rank) {
        return this.rank === other.rank && this.rankClass === other.rankClass;
    }

    public greaterThan(other: Rank) {
        if (this.rankClass === other.rankClass) {
            return this.rankClass === RankClass.KYU ? this.rank < other.rank : this.rank > other.rank;
        }
        if (this.rankClass === RankClass.DAN) {
            return true;
        }
        return false;
    }
}

export function formatRank(rating: number) {
    const rank = new Rank(rating);
    return rank.format();
}

export function formatRankShort(rating: number) {
    const rank = new Rank(rating);
    return rank.formatShort();
}
