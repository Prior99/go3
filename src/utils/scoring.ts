export enum GameResult {
    WIN = "win",
    LOSS = "loss",
    TIE = "tie",
}

export enum RankClass {
    KYU = "kyu",
    DAN = "dan",
}

export interface Rank {
    rank: number;
    rankClass: RankClass;
}

/**
 * Implementation of the official EGF rating system.
 * http://www.europeangodatabase.eu/EGD/EGF_rating_system.php
 */
const epsilon = 0.016;

function winningExpectancy(ratingDifference: number) {
    // SE(A) = 1 / (e ^ (D / a) + 1) - epsilon / 2
    return 1 / (Math.exp(ratingDifference / adjustment(ratingDifference)) + 1) - epsilon / 2;
}

function adjustment(rating: number) {
    return 205 - Math.floor(rating / 20);
}

function magnitude(rating: number) {
    return Math.round(122 - 6 * (rating / 100) + Math.pow(rating / 100, 2) / 15);
}

function scoreFromResult(result: GameResult) {
    switch (result) {
        case GameResult.WIN: return 1;
        case GameResult.TIE: return 0.5;
        case GameResult.LOSS: return 0;
        default: return;
    }
}

export function newRating(oldRating: number, otherRating: number, result: GameResult) {
    const actualScore = scoreFromResult(result);
    const expectedScore = winningExpectancy(otherRating - oldRating);
    let difference = actualScore - expectedScore;
    return Math.round(oldRating + magnitude(oldRating) * (actualScore - expectedScore));
}

export function rankFromRating(rating: number): Rank {
    if (rating < 2100) {
        return {
            rank: Math.round((2000 - rating) / 100 + 1),
            rankClass: RankClass.KYU,
        };
    }
    return {
        rank: Math.round((rating - 2000) / 100),
        rankClass: RankClass.DAN,
    };
}

export function formatRank(rating: number) {
    const rank = rankFromRating(rating);
    if (rank.rankClass === RankClass.KYU) {
        return `${rank.rank} Kyu`;
    }
    return `${rank.rank} Dan`;
}

export function formatRankShort(rating: number) {
    const rank = rankFromRating(rating);
    if (rank.rankClass === RankClass.KYU) {
        return `${rank.rank}k`;
    }
    return `${rank.rank}d`;
}
