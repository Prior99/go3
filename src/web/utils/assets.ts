import { component } from "tsdi";
import { bindAll } from "lodash-decorators";
import * as downscale from "downscale";

function getKey(url: string, width?: number, height?: number) {
    if (typeof width !== "undefined" && typeof height !== "undefined") {
        return `${width}__${height}__${url}`;
    }
    return url;
}

@component
@bindAll()
export class Assets {
    private images = new Map<string, HTMLImageElement>();
    private loading = new Map<string, ((HTMLImageElement) => void)[]>();
    private scaling = new Map<string, ((HTMLImageElement) => void)[]>();

    public get loaded() { return this.loading.size === 0; }

    public loadImage(url: string, width?: number, height?: number) {
        if (typeof width === "undefined" || typeof height === "undefined") {
            return this.downloadImage(url);
        }
        return new Promise<HTMLImageElement>(async resolve => {
            const key = getKey(url, width, height);
            if (this.images.has(key)) {
                return resolve(this.images.get(key));
            }
            if (!this.scaling.has(key)) {
                this.scaling.set(key, []);
                const original = await this.downloadImage(url);
                const scaled = await downscale(original, width, height, { imageType: "png" });
                const scaledImg = new Image();
                scaledImg.addEventListener("load", () => {
                    this.images.set(key, scaledImg);
                    this.scaling.get(key).forEach(callback => callback(scaledImg));
                    this.scaling.delete(key);
                });
                scaledImg.src = scaled;
            }
            this.scaling.get(key).push(resolve);
        });
    }

    private downloadImage(url: string) {
        return new Promise<HTMLImageElement>(resolve => {
            if (this.images.has(url)) {
                return resolve(this.images.get(url));
            }
            if (!this.loading.has(url)) {
                this.loading.set(url, []);
                const img = new Image();
                img.addEventListener("load", () => {
                    this.images.set(url, img);
                    this.loading.get(url).forEach(callback => callback(img));
                    this.loading.delete(url);
                });
                img.src = url;
            }
            this.loading.get(url).push(resolve);
        });
    }

    public get(url: string, width?: number, height?: number) {
        return this.images.get(getKey(url, width, height));
    }
}
