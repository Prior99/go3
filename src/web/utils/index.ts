export * from "./assets";
export * from "./rendering";
