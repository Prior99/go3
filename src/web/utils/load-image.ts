import { component } from "tsdi";
import { bind } from "decko";

@component
export class Assets {
    private images = new Map<string, HTMLImageElement>();
    private loading = new Map<string, ((HTMLImageElement) => void)[]>;

    public get loaded() { return loading.length === 0; }

    @bind public loadImage(url: string) {
        return new Promise<HTMLImageElement>(resolve => {
            if (!this.loading.has(url)) {
                this.loading.set(url, []);
            }
            this.loading.get(url).push(resolve);
            const img = new Image();
            img.src = url;
            img.addEventListener("load", () => {
                this.images.set(url, img);
                this.loading.get(url).forEach(callback => callback(img));
                this.loading.delete(url);
            });
        });
    }

    @bind public get(url: string) {
        return this.images.get(url);
    }
}
