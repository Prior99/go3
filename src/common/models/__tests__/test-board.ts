import { Board } from "..";
import { Color } from "../../utils";

function testBoard(test: string): Board {
    const state = [];
    for (let i = 0; i < test.length; ++i) {
        const character = test.charAt(i);
        const color =
            character === "W" ? Color.WHITE :
            character === "B" ? Color.BLACK :
            Color.EMPTY;
        state.push(color);
    }
    const board = new Board();
    board.state = state;
    return board;
}

test("Detects a hole", () => {
    const board = testBoard(
        "WWWW" +
        "W  W" +
        "W  W" +
        "WWWW",
    );
    expect(board.getEyeGroups(board.groupAt(0))).toMatchSnapshot();
});
