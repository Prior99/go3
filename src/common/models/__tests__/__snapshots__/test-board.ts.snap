// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Detects a hole 1`] = `undefined`;
